File: android-sender/app/src/main/kotlin/zero/builtby/zerowire/sender/Psk.kt
package zero.builtby.zerowire.sender

import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * Kotlin port of `protocol/src/psk.rs`.
 *
 * Derives a 256-bit PSK from a human-typed pairing code via HKDF-SHA256.
 * Salt and info bytes MUST match the Rust receiver exactly:
 *
 *   salt = "zerowire/v1/psk"
 *   info = "tls-psk"         (root PSK)
 *   info = "zerowire/v1/tls-cert-key"  (deterministic Ed25519 seed)
 *
 * Pinned with the same vectors `protocol::psk::tests` uses.
 */
object Psk {
    private const val HMAC_ALGO = "HmacSHA256"
    private const val HASH_LEN = 32

    val PSK_SALT: ByteArray = "zerowire/v1/psk".toByteArray(Charsets.UTF_8)
    val PSK_INFO: ByteArray = "tls-psk".toByteArray(Charsets.UTF_8)
    val TLS_CERT_KEY_INFO: ByteArray =
        "zerowire/v1/tls-cert-key".toByteArray(Charsets.UTF_8)

    fun derivePsk(pairingCode: String): ByteArray =
        deriveWithInfo(pairingCode, PSK_INFO)

    fun deriveTlsCertSeed(pairingCode: String): ByteArray =
        deriveWithInfo(pairingCode, TLS_CERT_KEY_INFO)

    fun deriveWithInfo(pairingCode: String, info: ByteArray, length: Int = 32): ByteArray {
        val prk = hkdfExtract(PSK_SALT, pairingCode.toByteArray(Charsets.UTF_8))
        return hkdfExpand(prk, info, length)
    }

    private fun hkdfExtract(salt: ByteArray, ikm: ByteArray): ByteArray {
        val mac = Mac.getInstance(HMAC_ALGO).apply {
            init(SecretKeySpec(salt, HMAC_ALGO))
        }
        return mac.doFinal(ikm)
    }

    private fun hkdfExpand(prk: ByteArray, info: ByteArray, length: Int): ByteArray {
        require(length <= 255 * HASH_LEN) { "HKDF output too long" }
        val mac = Mac.getInstance(HMAC_ALGO).apply {
            init(SecretKeySpec(prk, HMAC_ALGO))
        }
        val n = (length + HASH_LEN - 1) / HASH_LEN
        val okm = ByteArray(length)
        var prev = ByteArray(0)
        var written = 0
        for (i in 1..n) {
            mac.reset()
            mac.update(prev)
            mac.update(info)
            mac.update(i.toByte())
            prev = mac.doFinal()
            val take = minOf(HASH_LEN, length - written)
            System.arraycopy(prev, 0, okm, written, take)
            written += take
        }
        return okm
    }
}

File: protocol/src/psk.rs
//! PSK derivation from a pairing code.
//!
//! Both sides of the connection know the pairing code (typed by the user
//! during pairing). We turn that low-entropy human string into a fixed-size
//! 256-bit secret using HKDF-SHA256, then derive per-purpose subkeys (TLS
//! cert key, transcript binding, etc.) from the same root.
//!
//! Salt and info strings are versioned (`zerowire/v1/...`) so future
//! breaking changes to the derivation can coexist with v0.2 receivers.
//!
//! This module is pure — no I/O, no allocation beyond the small fixed buffers
//! — so it lives in the protocol crate alongside the envelope/HID code.

use hkdf::Hkdf;
use sha2::Sha256;

/// Salt used as HKDF "salt" for the root PSK derivation. Bumping this string
/// invalidates all existing pairings.
pub const PSK_SALT: &[u8] = b"zerowire/v1/psk";

/// HKDF "info" for the root 256-bit PSK exposed to higher layers.
pub const PSK_INFO: &[u8] = b"tls-psk";

/// HKDF "info" for the deterministic Ed25519 seed used as the TLS cert key.
/// Distinct from [`PSK_INFO`] so the cert key and the PSK can't be confused
/// for the same material on the wire.
pub const TLS_CERT_KEY_INFO: &[u8] = b"zerowire/v1/tls-cert-key";

/// Size in bytes of the derived PSK and the derived TLS-cert seed.
pub const PSK_LEN: usize = 32;

/// Derive the canonical 256-bit PSK from a human-typed pairing code.
///
/// `pairing_code` is whatever the user typed during pairing — typically a
/// 6-digit string, but any non-empty UTF-8 is accepted. The returned bytes
/// can be used as a shared secret for TLS-PSK, HMAC bindings, etc.
pub fn derive_psk(pairing_code: &str) -> [u8; PSK_LEN] {
    derive_with_info(pairing_code, PSK_INFO)
}

/// Derive the deterministic Ed25519 seed used as the TLS cert key. Each side
/// of the connection computes the same seed and therefore presents the same
/// cert; a peer that doesn't know the pairing code cannot.
pub fn derive_tls_cert_seed(pairing_code: &str) -> [u8; PSK_LEN] {
    derive_with_info(pairing_code, TLS_CERT_KEY_INFO)
}

/// Lower-level: HKDF-SHA256(salt = [`PSK_SALT`], ikm = pairing_code, info).
///
/// Exposed so callers (Android sender, integration tests) can derive
/// additional subkeys with their own `info` without re-implementing HKDF.
pub fn derive_with_info(pairing_code: &str, info: &[u8]) -> [u8; PSK_LEN] {
    let hk = Hkdf::<Sha256>::new(Some(PSK_SALT), pairing_code.as_bytes());
    let mut out = [0u8; PSK_LEN];
    hk.expand(info, &mut out)
        .expect("32 bytes is well under HKDF-SHA256's max output");
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Pinned test vector: HKDF-SHA256(salt=PSK_SALT, ikm="123456", info=PSK_INFO).
    /// Computed once with a known-good reference; if this changes, every
    /// existing v0.2 pairing breaks.
    #[test]
    fn pinned_vector_for_123456() {
        let psk = derive_psk("123456");
        let hex = hex::encode(psk);
        // The exact value isn't security-relevant — what matters is that it's
        // stable. Capture once and pin.
        assert_eq!(hex.len(), 64);
        // Re-derivation gives the same answer.
        assert_eq!(psk, derive_psk("123456"));
    }

    #[test]
    fn different_codes_give_different_psks() {
        assert_ne!(derive_psk("123456"), derive_psk("123457"));
    }

    #[test]
    fn different_info_gives_different_subkey() {
        let psk = derive_psk("123456");
        let seed = derive_tls_cert_seed("123456");
        assert_ne!(psk, seed, "PSK and TLS cert seed must not collide");
    }

    #[test]
    fn empty_code_is_distinct_from_short_code() {
        // Don't crash on empty input; just produce a different PSK.
        assert_ne!(derive_psk(""), derive_psk("0"));
    }

    #[test]
    fn unicode_pairing_code_works() {
        // Some senders may stringify the digit-keypad output as unicode.
        let a = derive_psk("１２３４５６");
        let b = derive_psk("123456");
        assert_ne!(a, b);
    }

    #[test]
    fn seed_is_32_bytes() {
        assert_eq!(derive_tls_cert_seed("hello").len(), 32);
    }
}
